### PR TITLE
Implement ICorProfilerInfo14::GetNonGCHeapBounds

### PR DESCRIPTION
**Pull request by EgorBo**
Contributes to https://github.com/dotnet/diagnostics/issues/4156

### CODE REVIEW

File: src/coreclr/vm/frozenobjectheap.cpp [hunk 1]
```diff
@@ -226,9 +226,7 @@ Object* FrozenObjectSegment::GetNextObject(Object* obj) const
     uint8_t* nextObj = (reinterpret_cast<uint8_t*>(obj) + ALIGN_UP(obj->GetSize(), DATA_ALIGNMENT));
     if (nextObj < m_pCurrent)
     {
-        Object* result = reinterpret_cast<Object*>(nextObj);
-        INDEBUG(result->Validate());
-        return result;
+        return reinterpret_cast<Object*>(nextObj);
```
Comment (EgorBo) [inline src/coreclr/vm/frozenobjectheap.cpp:229]: we agreed not to use COOP (`Validate()` needs it)

File: src/tests/profiler/native/nongcheap/nongcheap.cpp [hunk 2]
```diff
@@ -72,8 +113,29 @@ HRESULT NonGcHeapProfiler::GarbageCollectionFinished()
...
+                {
+                    printf("EnumerateNonGCObjects: firstObj != obj\n!");
+                    _failures++;
+                }
+            }
+
+            // Add test coverage for IsFrozenObject API, currently, it is expected to return true
+            // for objects from non-GC heap (it might also return true for frozen segments we don't track)
+            BOOL isFrozen;
+            hr = pCorProfilerInfo->IsFrozenObject(obj, &isFrozen);
```
Comment (EgorBo) [inline src/tests/profiler/native/nongcheap/nongcheap.cpp:131]: @cshung added test coverage for `IsFrozenObject` on your request

**Comment by EgorBo** [comment]
Failures are https://github.com/dotnet/runtime/issues/85081 and ci timeouts

**Comment by EgorBo** [comment]
@davmason @noahfalk PTAL

File: src/coreclr/vm/proftoeeinterfaceimpl.cpp [hunk 3]
```diff
@@ -7624,6 +7628,60 @@ HRESULT ProfToEEInterfaceImpl::EnumerateNonGCObjects(ICorProfilerObjectEnum** pp
...
+            ranges[segIdx].rangeStart = (ObjectID)firstObj;
+
+            // Total size reserved for a segment
+            ranges[segIdx].rangeLengthReserved = (UINT_PTR)segments[segIdx]->m_Size;
+
+            // Size of the segment that is currently in use
+            ranges[segIdx].rangeLength = (UINT_PTR)(segments[segIdx]->m_pCurrent - firstObj);
+        }
+
+        *pcObjectRanges = segmentsToInspect;
```
Comment (davmason) [inline src/coreclr/vm/proftoeeinterfaceimpl.cpp:7676]: For the other APIs we allow people to pass NULL for the count argument, can you add a null check around the assignments for `pcObjectRanges`?

**Review by davmason — APPROVED** [review]
Just the one comment, otherwise looks good

**Comment by EgorBo** [comment]
Failures are known, https://github.com/dotnet/runtime/issues/85637

File: src/coreclr/vm/proftoeeinterfaceimpl.cpp [hunk 4]
```diff
@@ -7624,6 +7628,66 @@ HRESULT ProfToEEInterfaceImpl::EnumerateNonGCObjects(ICorProfilerObjectEnum** pp
...
+            // Total size reserved for a segment
+            ranges[segIdx].rangeLengthReserved = (UINT_PTR)segments[segIdx]->m_Size;
+
+            // Size of the segment that is currently in use
+            ranges[segIdx].rangeLength = (UINT_PTR)(segments[segIdx]->m_pCurrent - firstObj);
+        }
+
+        if (pcObjectRanges != nullptr)
+        {
+            *pcObjectRanges = segmentsToInspect;
```
Comment (noahfalk) [inline src/coreclr/vm/proftoeeinterfaceimpl.cpp:7678]: @EgorBo - I think we want `*pcObjectRanges = segmentsCount`

I'd expect the current behavior will make a test case like this fail:
```
ULONG count
GetNonGCHeapBounds(0, &count, nullptr);
ranges = new ranges[count];
GetNonGCHeapBounds(count, nullptr, ranges);
```

Comment (EgorBo) [reply src/coreclr/vm/proftoeeinterfaceimpl.cpp:7678]: > @EgorBo - I think we want `*pcObjectRanges = segmentsCount`
> 
> I'd expect the current behavior will make a test case like this fail:
> 
> ```
> ULONG count
> GetNonGCHeapBounds(0, &count, nullptr);
> ranges = new ranges[count];
> GetNonGCHeapBounds(count, nullptr, ranges);
> ```

Ah, makes sense, I thought it was "how many entries were written to ranges". Will fix

Comment (noahfalk) [reply src/coreclr/vm/proftoeeinterfaceimpl.cpp:7678]: Yeah, I've seen it used that way too in other interfaces, but I checked GetGenerationBounds() for the GC case and it appears to have the segmentsCount semantics there so presumably we want to match the behavior.

**Pull request merged**